### ISSUE TITLE
False-positive annotation when returning Self::Type

### ISSUE THREAD

**Issue by axos88** [intellij-rust/intellij-rust#3029]
## Environment

* **Intellij-Rust plugin version**: 0.2.0.2103-182
* **Rust toolchain version**: rustc 1.28.0 (9634041f0 2018-07-30)
* **IDE name and version**: IntelliJ IDEA 2018.2 (Ultimate Edition) Build #IU-182.3684.101, built on July 24, 2018
* **Operating system**: OS X 10.12.6

## Problem description

Syntax highligh error: ![image](https://user-images.githubusercontent.com/1281218/48313203-2e1a8300-e5b9-11e8-89d4-266f91d4a2b7.png)

May be related to #1549

## Steps to reproduce

```rust
trait Y {
  type Result;
  fn handle() -> Result
}

impl Y for X {
    type Result = ();

    fn handle() -> Self::Result {
      debug!("Foo");
    }
}
```

**Comment by Undin** [comment]
@axos88 On the one hand, I can confirm that we have false positive error annotation for your code
On the other hand, your code is incomplete.
Just add a definition of `X` and wrong error annotation will go away

**Comment by axos88** [comment]
Heh, sorry. I will add a complete code later. This is actually happening for code that compiles correctly, so I'm sure the problem is not that is incomplete.

Happens with actix actors, btw.

**Comment by axos88** [comment]
![image](https://user-images.githubusercontent.com/1281218/48344899-b6129280-e676-11e8-9b78-e6499b003bd4.png)

Same with X defined
